overlord/configstate/config/transaction_test.go

```diff
@@ -604,3 +606,229 @@ func (s *transactionSuite) TestPristineGet(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(res2, Equals, "other-value")
 }
+
+func (s *transactionSuite) TestVirtualGetError(c *C) {
+
+	tests := []string{
+		"/", "..", "ä#!",
+		"a..b",
+	}
+
+	for _, tc := range tests {
+		err := config.RegisterVirtualConfig("some-snap", tc, func(key string) (interface{}, error) {
+			return nil, nil
+		})
+		c.Assert(err, ErrorMatches, "cannot register virtual config: invalid option name:.*")
+	}
+}
+
+func (s *transactionSuite) TestVirtualGetSimple(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+	s.state.Set("config", map[string]map[string]interface{}{
+		"some-snap": {
+			"other-key": "other-value",
+		},
+	})
+
+	n := 0
+	err := config.RegisterVirtualConfig("some-snap", "key.virtual", func(key string) (interface{}, error) {
+		n++
+
+		s := fmt.Sprintf("%s=virtual-value", key)
+		return s, nil
+	})
+	c.Assert(err, IsNil)
+
+	tr := config.NewTransaction(s.state)
+
+	var res string
+	// non-virtual keys work fine
+	err = tr.Get("some-snap", "other-key", &res)
+	c.Assert(err, IsNil)
+	c.Check(res, Equals, "other-value")
+	// no virtual helper was called because the requested key was not
+	// part of the virtual configuration
+	c.Check(n, Equals, 0)
+
+	// simple case: subkey is virtual
+	err = tr.Get("some-snap", "key.virtual", &res)
+	c.Assert(err, IsNil)
+	c.Check(res, Equals, "key.virtual=virtual-value")
+	// the virtual config function was called now
+	c.Check(n, Equals, 1)
+}
+
+func (s *transactionSuite) TestVirtualDeepNesting(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	config.RegisterVirtualConfig("some-snap", "key.virtual", func(key string) (interface{}, error) {
+		c.Check(key, Equals, "key.virtual")
+
+		m := make(map[string]string)
+		m["subkey"] = fmt.Sprintf("nested-value")
+		m["other-subkey"] = fmt.Sprintf("other-nested-value")
+
+		return m, nil
+	})
+
+	tr := config.NewTransaction(s.state)
+	var res string
+	err := tr.Get("some-snap", "key.virtual.subkey", &res)
+	c.Assert(err, IsNil)
+	c.Check(res, Equals, "nested-value")
+}
+
+func (s *transactionSuite) TestVirtualSetShadowsVirtual(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	err := config.RegisterVirtualConfig("some-snap", "key.nested.virtual", func(key string) (interface{}, error) {
+		c.Fatalf("unexpected cal to virtual config function")
+		return nil, nil
+	})
+	c.Assert(err, IsNil)
+
+	tests := []struct {
+		snap, key, value string
```

Comment:
minor: it might also be nice to test the case when `value` is a map.

Reply:
Nice idea, thank you! added